lib: Handle metadata changes for send-only folders (fixes #4616, fixes #4627)

imsodin opened the pull request:
As promised, this is the followup to #4584, fixing two more bugs. There are two main changes in behaviour:

The first one is simple: Send-only folders gain a "puller", which is triggered the same way as for the send-receive folder. It checks needed files for ignored files and files, that only differ by internal metadata.  

The second one is about unignoring files: Before this PR when a file is unignored, during scanning its version is updated. I.e. the unignored, potentially vastly outdated compared to the global version file is propagated to all other devices. The new behaviour is, that such a file is moved into conflict with the current global file. This is achieved by removing all counters except our own (which was increased) in the version. This obviously has the potential to create a lot of conflict files, but I consider this vastly preferable to silently overwriting the current global state with the previously ignored file.

calmh commented:
I haven't read this yet but it sounds good.

Can it also fix the case where two devices are initialized with the same set of files and one is send only? Currently it'll end up about 50% out of sync, but the send only side could accept changes that only affect the version vector (I.e., not file data or metadata) and it would just work.

imsodin commented:
Yes it can - I guess this is the main usage scenario next to "both devices delete a file" described in the issue.

The current implementation has one downside I didn't mention above: All out-of-sync files in a send-only folder are looked up in the database whenever an index update comes in. This could be improved, if on index update only the "new" files are looked up, but afaik there is no way to distinguish these from "old" ones. I don't think this is very bad though compared to what it gains.

One more thing: The condition determining whether two files are equal is aweful:
```
curFile.Deleted != intf.IsDeleted() || (!curFile.Deleted && (file.Size != curFile.Size || file.Type != curFile.Type || (!f.ignorePermissions(file) && file.Permissions != curFile.Permissions) || !file.ModTime().Equal(curFile.ModTime()) || !blocksEqual(file.Blocks, curFile.Blocks)))
```
Similar conditions exist in walk.go. I was considering consolidating all of these to the protocol package in the form of a `func (f FileInfo) IsEqual(other FileInfo) bool` function - does that sound reasonable?

calmh commented on lib/db/leveldb_dbinstance.go:
```diff
@@ -351,12 +351,6 @@ func (db *Instance) withNeed(folder, device []byte, truncate bool, needAllInvali
 			if bytes.Equal(v.Device, device) {
 				have = true
 				haveFileVersion = v
-				// We need invalid files regardless of version when
```
What's the thing behind the need invalid stuff that we just added now no longer making sense? Just a couple of sentences of background for my edification?

(It's a minus diff and I was confused when it was added so I'm not complaining, it's just that I'm now confused again.)

imsodin replied:
Before the invalidation of ignored files was introduced, `WithNeed` always returned ignored files, as it's version wasn't updated. Their version was updated to keep "in-sync" with the global version. Therefore these files would not get pulled unless the global version gets updated. The removed option returned all invalid files and was triggered when ignore patterns change. With the new mechanism of setting the version to conflict with the global state when unignoring, this is no longer needed.

calmh commented on lib/model/folder.go:
```diff
@@ -238,3 +251,25 @@ func (f *folder) setError(err error) {
+// set on the local host or the FlagNoPermBits has been set on the file/dir
+// which is being pulled.
+func (f *folder) ignorePermissions(file protocol.FileInfo) bool {
+	return f.IgnorePerms || file.NoPermissions
```
Micro-nit: I'm not sure this actually merits a helper function. The code is shorter and clearer than the comment.

calmh commented on lib/model/requests_test.go:
```diff
@@ -286,17 +284,159 @@ func TestRequestVersioningSymlinkAttack(t *testing.T) {
 	}
+}
+
+func TestPullInvalidIgnoredSO(t *testing.T) {
```
Nano-nit: put these above the helper function, so the helper function makes sense when encountered. (I had to cmd-f to see if it was actually called).

calmh commented on lib/model/requests_test.go:
```diff
@@ -286,17 +284,159 @@ func TestRequestVersioningSymlinkAttack(t *testing.T) {
-	if err != nil {
-		panic("Failed to create temporary testing dir")
+// This test checks that (un-)ignored/invalid/deleted files are treated as expected.
+func pullInvalidIgnored(t *testing.T, ft config.FolderType) {
```
t.Helper() somewhere here

calmh commented on lib/model/rofolder.go:
```diff
@@ -70,3 +75,62 @@ func (f *sendOnlyFolder) String() string {
+
+		if ignores.ShouldIgnore(intf.FileName()) {
+			file := intf.(protocol.FileInfo)
+			file.Invalidate(f.model.id.Short())
```
(We have this and `f.model.shortID` sprinkled both here and in rwfolder so not blaming you, but in both cases I think we should just get a `myID` in the constructor and remember it instead of reaching into the model... Doesn't need to be fixed in this PR, just mentioning it so it's out there and maybe someone remembers to clean it up)

imsodin replied:
Done anyway (the `f.model.id.Short()` thing is likely entirely my product). I also took the opportunity to structure the folder members by function.

calmh commented on lib/model/rofolder.go:
```diff
@@ -70,3 +75,62 @@ func (f *sendOnlyFolder) String() string {
+		}
+
+		file := intf.(protocol.FileInfo)
+		if curFile.Deleted != intf.IsDeleted() || (!curFile.Deleted && (file.Size != curFile.Size || file.Type != curFile.Type || (!f.ignorePermissions(file) && file.Permissions != curFile.Permissions) || !file.ModTime().Equal(curFile.ModTime()) || !blocksEqual(file.Blocks, curFile.Blocks))) {
```
*This*, on the other hand, *does* merit a helper function!

imsodin replied:
Yep. I proposed in my last comment to move this check and the similar checks in walk.go to the protocol package into a `func (f FileInfo) IsEqual(other FileInfo) bool` function - does that sound like the right place?

imsodin replied:
@calmh I went ahead with it, [see this diff](https://github.com/syncthing/syncthing/pull/4750/files/f3409d3a4d05aa5577b8eef638bf8fe6b2ec7727..d9adcd584ffc983a2acfd607b88cb2c23452112f).

calmh commented on lib/protocol/bep_extensions.go:
```diff
@@ -126,6 +127,72 @@ func (f FileInfo) IsEmpty() bool {
+// A symlink is not "equal", if it has different
+//  - target
+// A directory does not have anything specific to check.
+func (f FileInfo) IsEqual(other FileInfo, ignorePerms bool, noBlocks bool) bool {
```
Params should either both be `ignore...` or `no...`, imho

calmh commented on lib/protocol/bep_extensions.go:
```diff
@@ -126,6 +127,72 @@ func (f FileInfo) IsEmpty() bool {
+// A directory does not have anything specific to check.
+func (f FileInfo) IsEqual(other FileInfo, ignorePerms bool, noBlocks bool) bool {
+	if f.IsEmpty() || other.IsEmpty() {
+		return f.IsEmpty() == other.IsEmpty()
```
With this definition, two fileinfos can be "equal" if they both have no version, even if everything else is different. That's odd and might bite us in the future.

calmh commented on lib/protocol/bep_extensions.go:
```diff
@@ -126,6 +127,72 @@ func (f FileInfo) IsEmpty() bool {
+	}
+
+	// characteristics common to all types
+	if f.Type != other.Type || f.Deleted != other.Deleted || f.Invalid != other.Invalid || (!(ignorePerms || f.NoPermissions || other.NoPermissions) && !PermsEqual(f.Permissions, other.Permissions)) {
```
Break this up a bit, especially the permissions juggling.

Also, we're considering fileinfos equal even if the name differs?

calmh commented on lib/protocol/bep_extensions.go:
```diff
@@ -126,6 +127,72 @@ func (f FileInfo) IsEmpty() bool {
+		return true
+	}
+
+	// Should the deprecates symlinks type be handled somehow?
```
no

calmh commented on lib/protocol/bep_extensions.go:
```diff
@@ -126,6 +127,72 @@ func (f FileInfo) IsEmpty() bool {
+// A file is not "equal", if it has different
+//  - modification time
+//  - size
+//  - blocks, unless there are no blocks to compare (scanning)
```
Why don't we care about the version vector? I'm generally suspicious of this definition of "IsEqual". IMHO, "IsEqual" means that every attribute on the struct is exactly identical and nothing else. Possibly we can get away with `ignorePerms` and `ignoreBlocks`, but nothing else should be skipped unless we explicitly say so...

imsodin replied:
I should have stuck with the original name `IsEquivalent` or maybe `Unchanged`. Version (and Sequence, potentially others) is purposely excluded. The job of this function is to identify file infos that represent the same underlying files (currently used for scanning and send-only resolution). The conditions have been take over from walk.go.

calmh commented on lib/protocol/bep_extensions.go:
```diff
@@ -126,6 +127,72 @@ func (f FileInfo) IsEmpty() bool {
+
+// BlocksEqual returns whether two slices of blocks are exactly the same hash
+// and index pair wise.
+func BlocksEqual(src, tgt []BlockInfo) bool {
```
This used to live in `scanner`, then someone refactored that to a private function in `model` for some reason, and now it's in `protocol`. We should stop moving things around...

imsodin replied:
Yeah - I just needed it here and I certainly wont dip into model from protocol (can't anyway because circular) and as I thus had to move it, protocol felt like the right place (it's operating on protocol.BlockInfo after all).

calmh commented:
Good stuff.
